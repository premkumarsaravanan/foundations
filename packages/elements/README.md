# Reapit Elements

A collection of React components and utilities for building apps for Reapit Marketplace

## Contributing

### Typescript modules

- All modules are added to the src folder.
- React components should live in their own folder within the components folder and be PascalCased.
- Utilities should live in their own folder within the utilities folder and be kebab-cased.
- All components and utils should export from an `index.ts(x)` at the root of their folder.
- All components should be properly tested and contain their tests within their folder, following the `__tests__` convention.
- All modules should export their own type definitions.
- `index.tsx` at the root of `src` is the entry point for the app. All modules should be exported from here eg `export * from './components/Input'`

### Styles

- All components should use vanilla (S)CSS classes (no modules) - refactor where necessary.
- Styles live in the styles folder in all cases.
- Styles export from `index.scss` at the root of the styles project, ensure any new files are `@import`ed here.

### Building and Publishing

- NOTED: THIS PROCESS WON'T BUMP THE PACKAGE VERSION AUTOMATICALLY FOR YOU

- When a PR is created, checks will run to make sure testcases have been passed, code have passes linter standard. If one of checks fail, the PR won't able to be merged, and require the sumbmitter to update his/her code again.
- Create a PR to merge develop. When the PR merged to develop, there will be a tag published that have a version based on version field on package.json file. If there were a tag that has the same tag name created, the old tag would be overridden by the new tag. Install them on other by edit package.json as `@reapit/elements:git+ssh:git@github.com:reapit/elements.git#{tag}`. eg `@reapit/elements: "git+ssh:git@github.com:reapit/elements.git#v0.5.4-beta"`, or commandline: `yarn add @reapit/elements@git+ssh:git@github.com:reapit/elements.git#v0.5.4-beta`
- To release a stable version of npm package, create a PR to merge master. When the PR was merged, npm package will be published to npm , a new release will be created automatically, and storybook assets will be deployed to GH-pages.

### To use the project

- You will need an NPM token to install the package - this should be added to the `.npmrc` file.
- `yarn add @reapit/elements`to your project - you can then import the modules with your chosen module system.
- You will need to globally install / add the CSS file from `@reapit/elements/dist/index.css`.

### Storybook

- All React components should have their own Storybook stories in their own folder using the `component.stories.tsx` convention.
- You should add as many variants of your component as is helpful for future devs - these are our live docs.
- All future component work for generic components should be "Storybook" first if possible.
- To run a local Storybook instance with a dev server run `yarn storybook`.
- To build and publish to Github Pages at https://reapit.github.io/elements/ run `yarn deploy-storybook` - this happens by default in a postpublish hook when you deploy to NPM.
