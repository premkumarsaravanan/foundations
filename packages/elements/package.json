{
  "name": "@reapit/elements",
  "version": "0.5.34",
  "description": "A collection of React components and utilities for building apps for Reapit Marketplace",
  "main": "dist/index.js",
  "umd:main": "dist/elements.umd.production.js",
  "module": "dist/elements.esm.js",
  "typings": "dist/index.d.ts",
  "files": [
    "dist"
  ],
  "scripts": {
    "test": "cross-env TZ=UTC jest --coverage --silent",
    "test-dev": "cross-env TZ=UTC jest --watch --verbose",
    "test:ci": "cross-env TZ=UTC jest  --coverage --silent --detectOpenHandles --forceExit",
    "prebuild": "rimraf dist",
    "build": "cross-env NODE_ENV=production tsdx build --format=cjs,esm,umd && rollup -c",
    "start": "cross-env NODE_ENV=development tsdx watch",
    "storybook": "start-storybook -p 6006",
    "build-storybook": "build-storybook -o out",
    "release:develop": "node scripts/releaseDevelop.js",
    "release:master": "node scripts/releaseMaster.js",
    "checkVersion": "node ./scripts/checkVersion.js"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/reapit/elements.git"
  },
  "author": "Will McVay <wmcvay@reapit.com>",
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/reapit/elements/issues"
  },
  "homepage": "https://github.com/reapit/elements#readme",
  "devDependencies": {
    "@babel/core": "^7.5.5",
    "@storybook/addon-actions": "^5.2.6",
    "@storybook/addon-docs": "^5.2.6",
    "@storybook/addon-info": "^5.2.6",
    "@storybook/addon-links": "^5.2.6",
    "@storybook/addon-storysource": "^5.2.6",
    "@storybook/addon-viewport": "^5.2.6",
    "@storybook/addons": "^5.2.6",
    "@storybook/cli": "^5.2.6",
    "@storybook/react": "^5.2.6",
    "@storybook/storybook-deployer": "^2.8.1",
    "@storybook/theming": "^5.2.8",
    "@types/storybook__react": "^4.0.2",
    "@types/papaparse": "^5.0.3",
    "awesome-typescript-loader": "^5.2.1",
    "babel-loader": "^8.0.6",
    "compare-versions": "^3.5.1",
    "css-loader": "^3.2.0",
    "mockdate": "^2.0.5",
    "node-sass": "^4.12.0",
    "prettier": "^1.18.2",
    "purgecss": "^1.3.0",
    "purify-css": "^1.2.5",
    "react-docgen-typescript-loader": "^3.3.0",
    "rimraf": "^2.7.0",
    "rollup": "1.27.13",
    "rollup-plugin-scss": "^1.0.2",
    "sass-loader": "^7.2.0",
    "style-loader": "^1.0.0",
    "ts-node": "^8.3.0",
    "tsdx": "^0.7.2",
    "tslib": "^1.10.0"
  },
  "dependencies": {
    "bulma": "^0.7.5",
    "formik": "^2.0.4",
    "hardtack": "^4.1.0",
    "himalaya": "^1.1.0",
    "jsonwebtoken": "^8.5.1",
    "papaparse": "^5.1.1",
    "pell": "^1.0.6",
    "prop-types": "^15.7.2",
    "react-datasheet": "^1.4.0",
    "react-datepicker": "^2.9.6",
    "react-google-map": "^3.1.1",
    "react-google-maps-loader": "^4.2.5",
    "react-icons": "^3.7.0",
    "react-table": "7.0.0-alpha.35",
    "styled-components": "^4.4.1"
  },
  "peerDependencies": {
    "react": "^16.11.0",
    "react-dom": "^16.11.0",
    "react-router": "^5.1.2",
    "react-router-dom": "^5.1.2"
  },
  "browserslist": [
    "> 1%",
    "last 2 versions",
    "not ie <= 11"
  ]
}
