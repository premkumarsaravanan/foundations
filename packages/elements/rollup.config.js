import scss from 'rollup-plugin-scss'
import purify from 'purify-css'

export default [
  {
    input: 'src/styles/index.scss',
    output: {
      name: 'index',
      format: 'es'
    },
    plugins: [
      scss({
        output: (styles, styleNodes) => {
          const content = ['./dist/elements.esm.js']
          const options = {
            output: 'dist/index.css',
            minify: true,
            whitelist: [
              '*react-datepicker*',
              '*mt*',
              '*mb*',
              '*ml*',
              '*mr*',
              '*mx*',
              '*my*',
              '*pt*',
              '*pb*',
              '*pl*',
              '*pr*',
              '*px*',
              '*py*',
              '*flex*',
              '*justify*',
              '*items-center*',
              '*pin*',
              '*absolute*',
              '*relative*',
              '*capitalize*',
              'mw-100'
            ]
            // Uncomment this line if you want to see the CSS purified from the package
            // rejected: true
          }

          purify(content, styles, options)
        }
      })
    ]
  }
]
