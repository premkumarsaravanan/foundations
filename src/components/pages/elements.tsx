import * as React from 'react'
import ErrorBoundary from '@/components/hocs/error-boundary'

const SwaggerPage: React.SFC = () => (
  <ErrorBoundary>
    <iframe style={{ border: 'none' }} src="https://reapit.github.io/elements/" width="100%" height="100%" />
  </ErrorBoundary>
)

export default SwaggerPage
