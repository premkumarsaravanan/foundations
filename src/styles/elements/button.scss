// This version works but we can't theme as I am importing as raw CSS
// .callToAction {
//   composes: button is-primary from '../../../node_modules/bulma/css/bulma.css';
// }

// Webpack alias does not work
// .callToAction {
//   composes: button is-primary is-large from '~bulma/bulma.css'
// }

// Both versions below throw an unknown variable error
// .callToAction {
//   composes: button is-primary is-large from '../../../node_modules/bulma/sass/elements/button.sass'
// }

// This is the ideal way to use Bulma - we are using the version with our vaiable overrides
// .callToAction {
//   composes: button is-primary is-large from '../vendor/bulma.scss';
// }