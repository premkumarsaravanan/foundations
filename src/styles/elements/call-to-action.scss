@import '../base/colors.scss';

.container {
  max-width: 520px;
}

.isCard {
  background-color: $white;
  border: 1px solid $grey-lighter;
  border-radius: 5px;
  padding: 20px;
  box-shadow: 0 2px 3px rgba(18, 38, 63, 0.1);
}

.isCenter {
  text-align: center;
  margin: 0 auto;

  h3 {
    justify-content: center;
    flex-direction: column;
  }

  span:first-of-type {
    margin-bottom: 20px;
  }
}

.title {
  font-size: 20px;
  color: $black;
  display: flex;
  align-items: center;
  font-weight: bold;
  margin-bottom: 20px;
  border-bottom: 1px solid $grey-lighter;
  padding-bottom: 20px;
}

.icon {
  font-size: 20px;
  display: inline-flex;
  width: 40px;
  height: 40px;
  justify-content: center;
  align-items: center;
  border-radius: 999px;
  margin-right: 10px;
  font-size: 24px;
}

.iconSuccess {
  color: $white;
  background-color: $green;
}

.iconDanger {
  color: $white;
  background-color: $red;
}

.message {
  margin-bottom: 20px;
  a {
    text-decoration: underline;
  }
}
