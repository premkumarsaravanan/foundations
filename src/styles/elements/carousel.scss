@import '../../styles/base/colors.scss';

$slide-height: 310px;

.container {
  height: $slide-height;
  margin: 30px 0 0 0;

  :global .slick-list,
  :global .slick-track {
    height: $slide-height;
  }

  // style next/prev nav button
  button {
    opacity: 0;
    position: absolute;
    top: 50%;
    width: 40px;
    height: 40px;
    transform: translateY(-50%);
    z-index: 2;
    background-color: $grey-lightest;
    border: none;
    font-size: 20px;
    display: flex !important;
    align-items: center;
    justify-content: center;
    transition: all 0.2s ease;
    cursor: pointer;
    outline: none;

    &:first-of-type {
      left: -10px;
    }
    &:last-of-type {
      right: -10px;
      svg {
        transform: rotate(180deg);
      }
    }
  }

  &:hover {
    button {
      opacity: 1;
      &:first-of-type {
        left: 0px;
      }
      &:last-of-type {
        right: 0px;
      }
    }
  }
}

.slide {
  height: $slide-height;
  outline: none;
  img {
    max-height: 100%;
    max-width: auto !important;
  }
}
