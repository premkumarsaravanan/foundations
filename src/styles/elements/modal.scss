.medium {
  max-width: 640px;
}

.small {
  max-width: 360px;
}
