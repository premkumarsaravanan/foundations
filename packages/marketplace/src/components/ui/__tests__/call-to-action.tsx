import * as React from 'react'
import { shallow } from 'enzyme'
import CallToAction, { CallToActionCardProps } from '../call-to-action'
import toJson from 'enzyme-to-json'
import { AcButton } from '@reapit/elements'

const props: CallToActionCardProps = {
  buttonText: 'My button',
  title: 'My title'
}

describe('CallToAction', () => {
  it('should match a snapshot', () => {
    expect(toJson(shallow(<CallToAction {...props} />))).toMatchSnapshot()
  })

  it('should allow custom className', () => {
    const wrapper = shallow(<CallToAction {...props} className="addition" />)
    expect(wrapper.find('.addition')).toHaveLength(1)
  })

  it('simulates onButtonClick event', () => {
    const mockButtonClick = jest.fn()
    const wrapper = shallow(<CallToAction {...props} onButtonClick={mockButtonClick} />)
    wrapper.find('Button').simulate('click')
    expect(mockButtonClick).toBeCalledTimes(1)
  })

  afterEach(() => {
    jest.resetAllMocks()
  })
})
