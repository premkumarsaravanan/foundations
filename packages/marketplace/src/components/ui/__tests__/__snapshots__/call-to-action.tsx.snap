// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CallToAction should match a snapshot 1`] = `
<div
  className=""
  data-test=""
>
  <Component
    className="mb-0"
    message="My title"
    type="success"
  />
  <Component
    flexColumn={true}
    hasBackground={true}
    hasPadding={true}
  >
    <p />
    <Component>
      <Component
        dataTest=""
        type="button"
        variant="primary"
      >
        My button
      </Component>
    </Component>
  </Component>
</div>
`;
